wordpress-mobile/WordPress-Android
# Add some tooling for code freeze

This PR adds some Fastlane lanes to automate a part of the code freeze steps. The main goal is to avoid mistakes with version names, codes, etc. while issuing the various commands.

The script needs the name of the version to freeze and:
 - Pulls the develop branch from GH.
 - Verifies that the version in the `develop` branch is an `alpha` version.
 - Reads the `version code` from the alpha one.
 - Calculates the various version names and codes.
At the end of this process, it displays the results for approval. If something is miscalculated, the user can abort the script.
If everything is ok and the user confirms the operation, the script:
- Pulls the branches from GH.
- Create the new release branch.
- Updates the version strings.
- Protects the release branch on GH.

The script works only for the standard scenario. Corner cases will be handled in the classic way (there's usually little value in making overcomplicated script to handle corner cases).

Given the checks it performs, it should be tested within a real-life environment in order to have it work (you can tinker with other cases to verify the checks it performs).

To test:
1. General setup
1.1 Copy the `env.example` file to `~/.wpandroid-env.default`.
1.2 Generate a new GitHub personal access token and fill the `GHHEHLPER_ACCESS` field with it
1.3 Run `bundle install``

2. Test environment setup
The following test instructions are related to the current state of the repository where 11.1 is in code freeze and they will simulate the next code freeze (11.2).

2.1 please tag 11.1 in your release/11.1 branch (you will remove the tag later).

3. Code Freeze lane test
3.1 Run `bundle exec fastlane code_freeze codefreeze_version:11.2`.
3.2 Verify that it gets the version rights.
3.3 Confirm the operation to have it continue.
3.4 Verify that the version in the `release/11.1` branch has been correctly updated.
3.5 Verify that the version in the `develop` branch has been correctly updated.
3.6 Verify that the `release/11.2` branch has been correctly created and updated.
3.7 Verify that the `release/11.2` has been pushed to GH and it's protected.
3.8 Verify that the `11.2` milestone on GH has the frozen tag (the snowflake symbol)  in the title.
3.9 Verify that the `wpandroid_prs_list_11.1_11.2.txt` file has been created in your home, with the merged PRs list.

4. Clean up steps
After the test, please clean up the environment by reverting the operations performed by the script
4.1 Remove the branch protection for `release/11.2` branch on GH.
4.2 Delete the `release/11.2` branch on both GH and you local copy.
4.3 Remove the frozen tag from the title of the `11.2` milestone on GH.
4.4 Delete last commit from your local `develop` branch.
4.5 Delete last commit from your local `release/11.1` branch.
